### PR TITLE
Division and unary operators for SMT checker

### PR DESCRIPTION
**Pull request by chriseth**
Depends on https://github.com/ethereum/solidity/pull/2993

Fixes https://github.com/ethereum/solidity/issues/3021

It turns out, z3 implements truncation for signed division differently. It seems we have to add our own division.

### CODE REVIEW

**Comment by chriseth** [comment]
Fixed signed division by turning it into unsigned division. This seems to be the best solution since using quantifiers seems to quickly reach the limits of an SMT solver and the other option, `define-fun` is actually just a macro for exactly what I did there.

**Comment by axic** [comment]
Please rebase this after merging #2993.

File: libsolidity/formal/SMTChecker.cpp [hunk 1]
```diff
@@ -695,6 +782,18 @@ smt::Expression SMTChecker::maxValue(IntegerType const& _t)
 smt::Expression SMTChecker::expr(Expression const& _e)
 {
 	if (!m_expressions.count(&_e))
+	{
+		m_errorReporter.warning(_e.location(), "Internal error: Expression undefined for SMT solver." );
```
Comment (axic) [inline libsolidity/formal/SMTChecker.cpp:786]: Shouldnt these be assertions?

File: libsolidity/formal/SMTChecker.cpp [hunk 2]
```diff
@@ -695,6 +782,18 @@ smt::Expression SMTChecker::maxValue(IntegerType const& _t)
...
+		createExpr(_e);
+	}
+	return m_expressions.at(&_e);
+}
+
+void SMTChecker::createExpr(Expression const& _e)
+{
+	if (m_expressions.count(&_e))
+		m_errorReporter.warning(_e.location(), "Internal error: Expression created twice in SMT solver." );
```
Comment (axic) [inline libsolidity/formal/SMTChecker.cpp:795]: Shouldnt these be assertions?

File: libsolidity/formal/SMTChecker.cpp (hunk 1 above)
Comment (chriseth) [reply libsolidity/formal/SMTChecker.cpp:786]: If we leave them as warnings, you can still use the SMT checker on some functions while element of other functions are not yet supported.

Comment (axic) [reply libsolidity/formal/SMTChecker.cpp:786]: What kind of internal error still ensures that the rest is unaffected? (Not sure when the above can happen.)

Comment (chriseth) [reply libsolidity/formal/SMTChecker.cpp:786]: Ah sorry, this is only due to a missing implementation for an AST node type.

**Comment by axic** [comment]
Build fails as it needs to be updated after the `sort` change.

File: test/libsolidity/SMTChecker.cpp [hunk 3]
```diff
@@ -458,6 +458,101 @@ BOOST_AUTO_TEST_CASE(for_loop)
...
+			}
+		}
+	)";
+	CHECK_SUCCESS_NO_WARNINGS(text);
+	text = R"(
+		contract C {
+			function f(int x, int y) public pure {
+				x = -7;
+				y = -2;
+				int r = x / y;
```
Comment (axic) [inline test/libsolidity/SMTChecker.cpp]: Why do you need a temporary variable?

Comment (chriseth) [reply test/libsolidity/SMTChecker.cpp]: We don't need one, I can remove it.

**Comment by chriseth** [comment]
Ready to merge?

**Review by axic — APPROVED** [review]

**Pull request merged**